src/Rings/mpoly-local.jl

```diff
@@ -0,0 +1,366 @@
+    r.Ox = parent(a[1])
+    r.Sx = singular_ring_loc(r.Ox, ord = ord)
+    return r
+  end
+end
+
+mutable struct MPolyIdeal_loc{S} <: Ideal{S}
+  gens::BiPolyArray_loc{S}
+  min_gens::BiPolyArray_loc{S}
+  gb::BiPolyArray_loc{S}
+  dim::Int
+
+  function MPolyIdeal_loc(Ox::T, s::Singular.sideal) where {T <: MPolyRing_loc}
+    r = new{elem_type(T)}()
+    r.dim = -1 #not known
+    r.gens = BiPolyArray_loc(Ox, s)
+    if s.isGB
+      r.gb = gens
+    end
+    return r
+  end
+  function MPolyIdeal_loc(B::BiPolyArray_loc{T}) where T
+    r = new{T}()
+    r.dim = -1
+    r.gens = B
+    return r
+  end
+  function MPolyIdeal_loc(g::Array{T, 1}) where {T <: MPolyElem_loc}
+    r = new{T}()
+    r.dim = -1 #not known
+    r.gens = BiPolyArray_loc(g)
+    return r
+  end
+end
+
+###############################################################################
+# Basic ideal functions                                                       #
+###############################################################################
+
+function Base.getindex(A::BiPolyArray_loc, ::Val{:S}, i::Int)
+  if !isdefined(A, :S)
+    A.S = Singular.Ideal(A.Sx, [convert(A.Sx, x) for x = A.O])
+  end
+  return A.S[i]
+end
+
+function Base.getindex(A::BiPolyArray_loc, ::Val{:O}, i::Int)
+  if !isassigned(A.O, i)
+    A.O[i] = convert(A.Ox, A.S[i])
+  end
+  return A.O[i]
+end
+
+function Base.length(A::BiPolyArray_loc)
+  if isdefined(A, :S)
+    return Singular.ngens(A.S)
+  else
+    return length(A.O)
+  end
+end
+
+function Base.iterate(A::BiPolyArray_loc, s::Int = 1)
+  if s > length(A)
+    return nothing
+  end
+  return A[Val(:O), s], s+1
+end
+
+Base.eltype(::BiPolyArray_loc{S}) where S = S
+
+function Base.show(io::IO, I::MPolyIdeal_loc)
+  print(io, "ideal generated by: ")
+  g = collect(I.gens)
+  first = true
+  for i = g
+    if first
+      print(io, i)
+      first = false
+    else
+      print(io, ", ", i)
+    end
+  end
+  print(io, "")
+end
+
+function Base.show(io::IO, ::IJuliaMime, I::MPolyIdeal_loc)
+  print(io, "\$")
+  math_html(io, I)
+  print(io, "\$")
+end
+
+function math_html(io::IO, I::MPolyIdeal_loc)
+  print(io, "\\text{ideal generated by: }")
+  g = collect(I.gens)
+  first = true
+  for i = g
+    if first
+      math_html(io, i)
+      first = false
+    else
+      print(io, ", ")
+      math_html(io, i)
+    end
+  end
+  print(io, "")
+end
+
+###############################################################################
+# Ideal constructor functions                                                 #
+###############################################################################
+
+function ideal(g::Array{T, 1}) where T <: MPolyElem_loc
+  return MPolyIdeal_loc(g)
+end
+
+function ideal(Rx::MPolyRing_loc, g::Array{<:Any, 1})
+  f = elem_type(Rx)[Rx(f) for f = g]
+  return ideal(f)
+end
+
+function ideal(Rx::MPolyRing_loc, g::Singular.sideal)
+  return MPolyIdeal_loc(Rx, g)
+end
+
+function singular_assure(I::MPolyIdeal_loc)
```

Comment:
The name of this function doesn't give a big hint as to what it does. I apologise if this is by now a standard thing, but my comment would still apply, I think.